### PR TITLE
Wer tracker

### PR DESCRIPTION
**Pull request by JorisCos**
# About this PR

- This PR  makes it possible to keep track of the transcriptions made by the ASR models.

The file containing the transcriptions is a `.json` that looks like this :
all_trans_ex

- This PR also adds jiwer transformation to the measure computation. It removes the punctuation and puts everything to lowercase. This leads to a more accurate WER.

### CODE REVIEW

File: asteroid/metrics.py [hunk 1]
```diff
@@ -146,6 +146,8 @@ def __init__(self, model_name, trans_df):
         self.input_txt_list = []
         self.clean_txt_list = []
         self.output_txt_list = []
+        self.transcription = []
```
Comment (mpariente) [inline asteroid/metrics.py:149]: Should be in plural right?

File: asteroid/metrics.py [hunk 2]
```diff
@@ -172,25 +174,41 @@ def __call__(
...
+            self.true_txt_list.append(dict(utt_id=tmp_id, text=self.trans_dic[tmp_id]))
+        # Mixture
         for tmp_id in wav_id:
             out_count = Counter(self.hsdi(truth=self.trans_dic[tmp_id], hypothesis=txt))
             self.mix_counter += out_count
             local_mix_counter += out_count
             self.input_txt_list.append(dict(utt_id=tmp_id, text=txt))
         # Average WER for the clean pair
-        for wav, tmp_id in zip(clean, wav_id):
+        for i, (wav, tmp_id) in enumerate(zip(clean, wav_id)):
```
Comment (mpariente) [inline asteroid/metrics.py:194]: That's a bit ugly and probably unclear for researchers. Not sure.

File: asteroid/metrics.py [hunk 3]
```diff
@@ -204,10 +222,17 @@ def wer_from_hsdi(hits=0, substitutions=0, deletions=0, insertions=0):
 
     @staticmethod
     def hsdi(truth, hypothesis):
+        import jiwer
         from jiwer import compute_measures
 
+        transformation = jiwer.Compose([jiwer.ToLowerCase(), jiwer.RemovePunctuation()])
```
Comment (mpariente) [inline asteroid/metrics.py:228]: Should this be defined in the constructor (init) instead?

**Review by mpariente — COMMENTED** [review]
The estimates seem a bit off :joy: but this was clearly missing, thanks!

File: asteroid/metrics.py (hunk 3 above)
Comment (JorisCos) [reply asteroid/metrics.py:228]: It should :)

**Comment by JorisCos** [comment]
Haha the estimates are wrong because the model was trained on noisy inputs and the enhancement is making things worse...

File: asteroid/metrics.py (hunk 3 above)
Comment (JorisCos) [reply asteroid/metrics.py:228]: Well, if we do that we have to make `self.transformation` an argument of `hsdi()` or `hsdi()` can stay a static method

Comment (mpariente) [reply asteroid/metrics.py:228]: Let's keep hsdi a static function and pass the transformation.

**Comment by mpariente** [comment]
Let's include the wav normalization in the WERTracker class?

**Comment by popcornell** [comment]
What is the ID field in .json annotation ?

**Comment by JorisCos** [comment]
> Let's include the wav normalization in the WERTracker class?

We can but we have to it anyway before saving the files.


> What is the ID field in .json annotation ?

In a general way the ID is something we introduced for librimix to match transcriptions and wav files . For this specific screen shot this annotation and ID's are taken from CHIME 4. ( I will open a PR soon)

**Comment by popcornell** [comment]
`In a general way the ID is something we introduced for librimix to match transcriptions and wav files . For this specific screen shot this annotation and ID's are taken from CHIME 4. ( I will open a PR soon)`

Maybe call em UtteranceID or ExampleID?  Because we might need also speaker IDs

**Comment by mpariente** [comment]
Also, please make the fields in JSON all lower case: "text_0", "utt_id_0" etc...

File: asteroid/metrics.py [hunk 4]
```diff
@@ -139,19 +139,23 @@ def __init__(self, model_name, trans_df):
...
         self.output_txt_list = []
+        self.transcriptions = []
+        self.true_txt_list = []
         self.sample_rate = int(d.data_frame[d.data_frame["name"] == model_name]["fs"])
         self.trans_df = trans_df
         self.trans_dic = self._df_to_dict(trans_df)
         self.mix_counter = Counter()
         self.clean_counter = Counter()
         self.est_counter = Counter()
+        self.transformation = jiwer.Compose([jiwer.ToLowerCase(), jiwer.RemovePunctuation()])
```
Comment (mpariente) [inline asteroid/metrics.py:158]: Is this transformation enough? 
The default is 
```
[<jiwer.transforms.RemoveMultipleSpaces at 0x7fbc79a75df0>,
 <jiwer.transforms.Strip at 0x7fbc79a75e20>,
 <jiwer.transforms.SentencesToListOfWords at 0x7fbc79a75f10>,
 <jiwer.transforms.RemoveEmptyStrings at 0x7fbc7aa17bb0>]
```

File: asteroid/metrics.py [hunk 5]
```diff
@@ -258,5 +295,8 @@ def final_df(self):
         )
         return df
 
+    def all_transcriptions(self):
+        return dict(transcriptions=self.transcriptions)
```
Comment (mpariente) [inline asteroid/metrics.py:299]: I don't really see the point of the dict with one field, returning the list. 
I'd remove this method entirely

File: asteroid/metrics.py (hunk 4 above)
Comment (JorisCos) [reply asteroid/metrics.py:158]: When I tested on CHIME4 these were the two that made a difference but you are right let's add the others. It doesn't cost that much anyway.

**Pull request merged**